File: src/main/java/me/rsman/BetterMinecraftCore/Managers/PlayerManager.java
package me.rsman.BetterMinecraftCore.Managers;

import me.rsman.BetterMinecraftCore.BetterMinecraftCore;
import me.rsman.BetterMinecraftCore.db.models.PlayerBaseAttr;
import me.rsman.BetterMinecraftCore.enums.EAttributes;
import org.bukkit.Material;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.ItemStack;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

public final class PlayerManager {
    public static Map<String, Map<String, Long>> playersAttributes = new HashMap<>();

    public static Map<String, Long> getBaseAttributes(String uuid){
        return PlayerManager.getBaseAttributes(uuid, false);
    }
    public static Map<String, Long> getBaseAttributes(String uuid,Boolean force){
        if (PlayerManager.playersAttributes.containsKey(uuid) && !force) {
            return PlayerManager.playersAttributes.get(uuid);
        }
        try {
            PlayerBaseAttr pba = DBManager.playerBaseAttrDao.queryForId(uuid);
            if(pba == null){
                DBManager.playerBaseAttrDao.create(new PlayerBaseAttr(uuid));
                pba = DBManager.playerBaseAttrDao.queryForId(uuid);
            }

            Map<String, Long> playerStats = new HashMap<>();
            for(Map.Entry<String, Long> attrs : pba.getMap().entrySet()){
                playerStats.put(attrs.getKey() + "_base", attrs.getValue());
                playerStats.put(attrs.getKey() + "_equip", 0L);
                playerStats.put(attrs.getKey() + "_skill", 0L);
                playerStats.put(attrs.getKey() + "_talisman", 0L);
            }
            PlayerManager.playersAttributes.put(uuid, playerStats);
            return playerStats;
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        return null;
    }

    public static void setBaseAttribute(String uuid, String attr, Long value){
        try{
            PlayerBaseAttr pba = DBManager.playerBaseAttrDao.queryForId(uuid);
            pba.setByName(attr, value);
            DBManager.playerBaseAttrDao.update(pba);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }

        PlayerManager.getBaseAttributes(uuid, true);
    }

    public static void setEquippedAttributes(String uuid, Map<String, Long>attr){
        Map<String,Long> attributes = PlayerManager.playersAttributes.get(uuid);
        for (Map.Entry<String, Long> entry : attr.entrySet()) {
            attributes.put(entry.getKey()+"_equip", entry.getValue());
        }
        PlayerManager.playersAttributes.put(uuid, attributes);
    }

    public static void setSkillAttributes(String uuid, Map<String, Long>attr){
        Map<String,Long> attributes = PlayerManager.playersAttributes.get(uuid);
        for (Map.Entry<String, Long> entry : attr.entrySet()) {
            attributes.put(entry.getKey()+"_skill", entry.getValue());
        }
        PlayerManager.playersAttributes.put(uuid, attributes);
    }

    public static void setTalismansAttributes(String uuid, Map<String, Long>attr){
        Map<String,Long> attributes = PlayerManager.playersAttributes.get(uuid);
        for (Map.Entry<String, Long> entry : attr.entrySet()) {
            attributes.put(entry.getKey()+"_talisman", entry.getValue());
        }
        PlayerManager.playersAttributes.put(uuid, attributes);
    }


    public static Map<String, Long> getAttributes(String uuid){
        Map<String,Long> returnAttributes = new HashMap<>();
        Map<String,Long> attributes = PlayerManager.playersAttributes.get(uuid);
        if(attributes == null){
            PlayerManager.getBaseAttributes(uuid, true);
        }
        attributes = PlayerManager.playersAttributes.get(uuid);
        for (String attr : EAttributes.getAllKeys()) {
            returnAttributes.put(attr,
                attributes.get(attr+"_base")+attributes.get(attr+"_equip")+attributes.get(attr+"_skill")+attributes.get(attr+"_talisman")
            );
        }
        return returnAttributes;
    }

    public static void alterPlayerAttributesWithEquippedStuff(Player player){
        BetterMinecraftCore.getInstance().getServer().getScheduler().runTaskLaterAsynchronously(BetterMinecraftCore.getInstance(),() -> {
            EntityEquipment equipment = player.getEquipment();
            assert equipment != null;
            ItemStack mainHand = equipment.getItemInMainHand();
            if(ItemTypeChecker.isArmorOrHead(mainHand)) mainHand = null;
            ItemStack offHand = equipment.getItemInOffHand();
            if(ItemTypeChecker.isArmorOrHead(offHand)) mainHand = null;
            ItemStack[] items = {
                    mainHand,
                    offHand,
                    equipment.getHelmet(),
                    equipment.getChestplate(),
                    equipment.getLeggings(),
                    equipment.getBoots(),
            };
            Map<String, Long> finalAttributes= new HashMap<>();
            for (String attr : EAttributes.getAllKeys()) { finalAttributes.put(attr, 0L); }
            for (ItemStack item : items) {
                if(item == null || item.getType() == Material.AIR || item.getType() == Material.WRITABLE_BOOK) continue;
                ItemManager.updateItem(item);
                for (String attr : EAttributes.getAllKeys()) {
                    long attrValue = ItemManager.getFinalItemAttr(item,attr);
                    finalAttributes.put(attr, finalAttributes.get(attr) + attrValue);
                }
            }
            setEquippedAttributes(player.getUniqueId().toString(), finalAttributes);
        },1);
    }

    public static void updatePlayerAttributes(Player player){
        String uuid = player.getUniqueId().toString();
        Map<String, Long> totalAttributes = PlayerManager.getAttributes(uuid);
        Map<String, Long> attributes = PlayerManager.playersAttributes.get(uuid);

        Long mana = attributes.get("currentMana");
        if(mana == null){
            attributes.put("currentMana", 0L);
            PlayerManager.playersAttributes.put(uuid, attributes);
            attributes = PlayerManager.playersAttributes.get(uuid);
            mana = attributes.get("currentMana");
        }
        mana = (long) Math.round(Math.min(Math.round(mana + totalAttributes.get("mana") * 0.02),totalAttributes.get("mana")));
        attributes.put("currentMana", mana);
        PlayerManager.playersAttributes.put(uuid, attributes);



        AttributeInstance playerAttackSpeed = player.getAttribute(Attribute.GENERIC_ATTACK_SPEED);
        if(playerAttackSpeed != null){
            playerAttackSpeed.setBaseValue(4*(1+ (double)totalAttributes.get("attackSpeed")/100));

        }
        AttributeInstance playerSpeed = player.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED);
        if(playerSpeed != null){
            playerSpeed.setBaseValue(0.1*(1+ (double)totalAttributes.get("speed")/100));
        }
        AttributeInstance playerMaxHealth = player.getAttribute(Attribute.GENERIC_MAX_HEALTH);
        if(playerMaxHealth != null){
            playerMaxHealth.setBaseValue(totalAttributes.get("health"));
        }

        player.setHealth(Math.min(totalAttributes.get("health"), player.getHealth()+(double)(totalAttributes.get("strength"))/200 + 1));
        player.setHealthScale(20);


        ActionBarManager.updateActionBar(player, totalAttributes, mana);
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/Managers/PlayerManager.kt
package me.rsman.BetterMinecraftCore.Managers

import java.util.HashMap
import me.rsman.BetterMinecraftCore.models.PlayerBaseAttr
import me.rsman.BetterMinecraftCore.BetterMinecraftCore
import me.rsman.BetterMinecraftCore.enums.EAttributes
import org.bukkit.entity.Player
import java.sql.SQLException
import org.bukkit.inventory.ItemStack
import java.lang.Runnable
import org.bukkit.Material
import org.bukkit.attribute.Attribute

object PlayerManager {
    var playersAttributes: MutableMap<String?, MutableMap<String, Long>> = HashMap()
    fun getBaseAttributes(uuid: String?): Map<String, Long>? {
        return getBaseAttributes(uuid, false)
    }

    fun getBaseAttributes(uuid: String?, force: Boolean?): Map<String, Long>? {
        if (playersAttributes.containsKey(uuid) && !force!!) {
            return playersAttributes[uuid]
        }
        try {
            var pba = DBManager.playerBaseAttrDao!!.queryForId(uuid)
            if (pba == null) {
                DBManager.playerBaseAttrDao!!.create(PlayerBaseAttr(uuid))
                pba = DBManager.playerBaseAttrDao!!.queryForId(uuid)
            }
            val playerStats: MutableMap<String, Long> = HashMap()
            for ((key, value) in pba!!.map) {
                playerStats[key.toString() + "_base"] = value ?: 0L
                playerStats[key.toString() + "_equip"] = 0L
                playerStats[key.toString() + "_skill"] = 0L
                playerStats[key.toString() + "_talisman"] = 0L
            }
            playersAttributes[uuid] = playerStats
            return playerStats
        } catch (throwables: SQLException) {
            throwables.printStackTrace()
        }
        return null
    }

    fun setBaseAttribute(uuid: String?, attr: String, value: Long) {
        try {
            val pba = DBManager.playerBaseAttrDao!!.queryForId(uuid)
            pba!!.setByName(attr, value)
            DBManager.playerBaseAttrDao!!.update(pba)
        } catch (throwables: SQLException) {
            throwables.printStackTrace()
        }
        getBaseAttributes(uuid, true)
    }

    fun setEquippedAttributes(uuid: String?, attr: Map<String, Long>) {
        val attributes = playersAttributes[uuid]!!
        for ((key, value) in attr) {
            attributes[key + "_equip"] = value
        }
        playersAttributes[uuid] = attributes
    }

    fun setSkillAttributes(uuid: String?, attr: Map<String, Long>) {
        val attributes = playersAttributes[uuid]!!
        for ((key, value) in attr) {
            attributes[key + "_skill"] = value
        }
        playersAttributes[uuid] = attributes
    }

    fun setTalismansAttributes(uuid: String?, attr: Map<String, Long>) {
        val attributes = playersAttributes[uuid]!!
        for ((key, value) in attr) {
            attributes[key + "_talisman"] = value
        }
        playersAttributes[uuid] = attributes
    }

    fun getAttributes(uuid: String?): Map<String, Long> {
        val returnAttributes: MutableMap<String, Long> = HashMap()
        var attributes: Map<String, Long>? = playersAttributes[uuid]
        if (attributes == null) {
            getBaseAttributes(uuid, true)
        }
        attributes = playersAttributes[uuid]
        for (attr in EAttributes.allKeys) {
            returnAttributes[attr] = attributes!![attr + "_base"]!! + attributes[attr + "_equip"]!! + attributes[attr + "_skill"]!! + attributes[attr + "_talisman"]!!
        }
        return returnAttributes
    }

    fun alterPlayerAttributesWithEquippedStuff(player: Player) {
        BetterMinecraftCore.instance.server.scheduler.runTaskLaterAsynchronously(BetterMinecraftCore.instance, Runnable {
            val equipment = player.equipment!!
            var mainHand: ItemStack? = equipment.itemInMainHand
            if (ItemTypeChecker.isArmorOrHead(mainHand)) mainHand = null
            val offHand = equipment.itemInOffHand
            if (ItemTypeChecker.isArmorOrHead(offHand)) mainHand = null
            val items = arrayOf(
                    mainHand,
                    offHand,
                    equipment.helmet,
                    equipment.chestplate,
                    equipment.leggings,
                    equipment.boots)
            val finalAttributes: MutableMap<String, Long> = HashMap()
            for (attr in EAttributes.allKeys) {
                finalAttributes[attr] = 0L
            }
            for (item in items) {
                if (item == null || item.type == Material.AIR || item.type == Material.WRITABLE_BOOK) continue
                ItemManager.updateItem(item)
                for (attr in EAttributes.allKeys) {
                    val attrValue = ItemManager.getFinalItemAttr(item, attr)
                    finalAttributes[attr] = finalAttributes[attr]!! + attrValue
                }
            }
            setEquippedAttributes(player.uniqueId.toString(), finalAttributes)
        }, 1)
    }

    fun updatePlayerAttributes(player: Player) {
        val uuid = player.uniqueId.toString()
        val totalAttributes = getAttributes(uuid)
        var attributes = playersAttributes[uuid]!!
        var mana = attributes["currentMana"]
        if (mana == null) {
            attributes["currentMana"] = 0L
            playersAttributes[uuid] = attributes
            attributes = playersAttributes[uuid]!!
            mana = attributes["currentMana"]
        }
        mana = Math.round(Math.min(Math.round(mana!! + totalAttributes["mana"]!! * 0.02), totalAttributes["mana"]!!).toFloat()).toLong()
        attributes["currentMana"] = mana
        playersAttributes[uuid] = attributes
        val playerAttackSpeed = player.getAttribute(Attribute.GENERIC_ATTACK_SPEED)
        if (playerAttackSpeed != null) {
            playerAttackSpeed.baseValue = 4 * (1 + (totalAttributes["attackSpeed"]?.toDouble() ?: 2.0) / 100)
        }
        val playerSpeed = player.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED)
        if (playerSpeed != null) {
            playerSpeed.baseValue = 0.1 * (1 + (totalAttributes["speed"]?.toDouble() ?: 0.0) / 100)
        }
        val playerMaxHealth = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)
        playerMaxHealth?.baseValue = totalAttributes["health"]?.toDouble() ?: 20.0
        player.health = (totalAttributes["health"]?.toDouble() ?: 20.0).coerceAtMost(player.health + (totalAttributes["strength"]?.toDouble() ?: 0.0) / 200 + 1)
        player.healthScale = 20.0
        ActionBarManager.updateActionBar(player, totalAttributes, mana)
    }
}